community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/expressions/Interpolation.scala

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2002-2015 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.compiler.v2_3.commands.expressions
+
+import org.neo4j.cypher.internal.compiler.v2_3._
+import org.neo4j.cypher.internal.compiler.v2_3.commands.values.InterpolationValue
+import org.neo4j.cypher.internal.compiler.v2_3.helpers.NonEmptyList
+import org.neo4j.cypher.internal.compiler.v2_3.pipes.QueryState
+import org.neo4j.cypher.internal.compiler.v2_3.symbols._
+
+import scala.annotation.tailrec
+import scala.collection.mutable
+
+case class Interpolation(parts: NonEmptyList[Either[Expression, String]]) extends Expression {
+  def apply(ctx: ExecutionContext)(implicit state: QueryState): Any = {
+    val partsValues = compile(parts, ctx, state, NonEmptyList.newBuilder[Either[String, String]])
+
+    partsValues.map(InterpolationValue).orNull
+  }
+
+  @tailrec
+  private def compile(
+    remaining: NonEmptyList[Either[Expression, String]],
+    ctx: ExecutionContext,
+    state: QueryState,
+    acc: mutable.Builder[Either[String, String], Option[NonEmptyList[Either[String, String]]]])
```

Comment:
Either[String, String] looks a bit weird, could it not just build up a `NonEmptyList[String]`? 
`InterpolationValue` doesn't seem to care if the string is coming from an expression or a String?